src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateChanges = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(stateClone, extraData);
+        break;
+      case 'removeProperties':
+        keysToRemove.forEach(key => delete stateClone[key]);
+        break;
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
```

Comment:
It would make more sense to also use `forEach` here. To start, get an array of keys using the `Object.keys` method.